### DISCUSSION TITLE
Can the Python packaging scan packages in all of the directories?

### DISCUSSION THREAD

**Discussion by ysk24ok** [aquasecurity/trivy#7315] [Q&A] [answered]
### Question

According to https://aquasecurity.github.io/trivy/v0.54/docs/coverage/language/python/, for Python Trivy has scans for
- packages managers
- packaging

My container has `requirements.txt` and packages installed from the file using `pip install -r`.
IIUC, [the pip scan](https://aquasecurity.github.io/trivy/v0.54/docs/coverage/language/python/#pip), which is one of the scans for package managers, checks packages listed in the file, and [the packaging scan](https://aquasecurity.github.io/trivy/v0.54/docs/coverage/language/python/#packaging) checks packages installed on the file system.

It's possible to install the packages to wherever I want, but does the packaging scan all of the directories?

### Target

Container Image

### Scanner

Vulnerability

### Output Format

None

### Mode

None

### Operating System

Linux

**Comment by knqyf263** [comment, marked as answer]
>does the packaging scan all of the directories?

Yes

**Reply by ysk24ok** [reply]
Thank you!

**Comment by subodh101** [comment]
@knqyf263 Can it also check for directories into a storage/PVC mounted in a pod?
If yes, that means trivy checks all the root level folders?
```
bin   dev  home  lib64	mnt  proc     root  sbin  sys  usr
boot  etc  lib	 media	opt  project  run   srv   tmp  var
```

**Reply by knqyf263** [reply]
Yes, some directories are skipped by default, though.
https://github.com/aquasecurity/trivy/blob/2b6d8d9227fb6ecc9386a14333964c23c0370a52/pkg/fanal/walker/walk.go#L17-L22

If you want, you can also pass `--skip-dirs`.

**Reply by subodh101** [reply]
Thank you for your reply.

https://github.com/aquasecurity/trivy/blob/d1f89672d92d04504808f9fa02845282c5e81b26/pkg/fanal/analyzer/language/python/pip/pip.go#L132
~Correct me if I am wrong with the flow of trivy execution?~
~1. If VIRTUAL_ENV is set, it tries to find the `lib` directory within the virtual environment and inside `lib` directory it looks for site-directory package.~
~2. From python executable’s directory, it constructs the path to `lib` directory by moving one level up and then followed by looking into site-directory package.~
~3. Lastly, It constructs a path two levels up from the python executable’s directory, then navigates to `lib/site-packages`.~

~And this is function keeps yielding the site-directory package paths to the parent function so that multiple location can be checked?~

~Why there is need to make a specific route because python executable’s directory could be located at a place deeper into the the folders such that constructing a path two levels up still won't find the `lib/site-packages`?~

~What happens if the python executable’s directory is not available in symbolic link or $PATH?~
Edit: Above not relevant to container image scanning.

**Reply by ysk24ok** [reply]
For scanning packages (egg or wheel), I think [python/packaging/packaging.go](https://github.com/aquasecurity/trivy/blob/main/pkg/fanal/analyzer/language/python/packaging/packaging.go) is used, not [python/pip/pip.go](https://github.com/aquasecurity/trivy/blob/main/pkg/fanal/analyzer/language/python/pip/pip.go), but I could be wrong.

**Reply by subodh101** [reply]
Okay, for "container scanning", only eggs and wheels are checked everywhere and skipping the some unnecessary folders?
https://aquasecurity.github.io/trivy/v0.54/docs/coverage/language/python/#packaging


[Not important] https://packaging.python.org/en/latest/discussions/wheel-vs-egg/ this link is broken, i found only a Japanese link.

**Reply by knqyf263** [reply]
>Okay, for "container scanning", only eggs and wheels are checked everywhere and skipping the some unnecessary folders?

Yes, as documented.

>[Not important] https://packaging.python.org/en/latest/discussions/wheel-vs-egg/ this link is broken, i found only a Japanese link.

We've fixed it in https://github.com/aquasecurity/trivy/pull/7318.
